#!/bin/env Rscript
################################################################################
#
# Performs Differential Expression Analysis using DESeq2
#
# Tests for differential expression between alternate binary partitionings
# of the data, generated by applying k-means clustering to alternate projections of
# the data + the original unprojected data.
#
# This step has no knowledge of the biological processes reflected in the data,
# but rather, is merely attempting to detect and characterize some of the most salient
# differences between samples.
#
# kh (mar.2020)
#
################################################################################
suppressMessages(library(arrow))
suppressMessages(library(DESeq2))
suppressMessages(library(tidyverse))

# load gene expression data
count_dat <- read_feather(snakemake@input[[1]]) %>%
  column_to_rownames("ensgene") %>%
  as.matrix()

# load sample cluster data
clusters <- read_feather(snakemake@input[[2]]) %>%
  column_to_rownames('sample_id')

# sanity check
if (!all(colnames(count_dat) == rownames(clusters))) {
  stop("Column mismatch for %s", snakemake@wildcards$project_id)
}

# drop any redundant clusterings
dups <- duplicated(t(clusters))
clusters <- clusters[, !duplicated(t(clusters)), drop=FALSE]

# convert binary cluster memberships to factors
for (i in 1:ncol(clusters)) {
  clusters[, i] <- factor(clusters[, i], labels = c('A', 'B'))
}

# iterate over clusterings and perform differential expression for each
for (clustering in colnames(clusters)) {
  dds <- DESeqDataSetFromMatrix(count_dat, clusters, as.formula(sprintf("~%s", clustering)))
  dds <- DESeq(dds)

  # first column correspond to the model intercept; second should be the cluster
  # comparison
  contrast <- resultsNames(dds)[2]

  # sanity check
  if (contrast == 'Intercept') {
    stop('Wrong contrast selected!')
  }

  res <- results(dds, name = contrast) %>%
    as.data.frame() %>%
    rownames_to_column("ensgene") %>%
    select(ensgene, everything())

  write_feather(res, snakemake@output[[1]])
}

